README.md
![build](https://github.com/transcriptaze/wav2png/workflows/build/badge.svg)

# wav2png

Renders a WAV file as a PNG image, with options to draw a grid, custom colouring and anti-aliasing. 

There are three implementations:

- A command line version (in the _go_ subdirectory), mostly intended for scripting and batch processing
- An online version implemented by compiling this library to WASM, which can be found
  [here](https://transcriptaze.github.io/W2P.html).
- A WebGPU version in the _webgpu_ directory) for a faster interactive experience, 
  hosted on [CloudFlare Pages](https://wav2png.pages.dev).

## Raison d'être

_wav2png_ was originally created as a Go utility library to render an audio file as an anti-aliased waveform for
a WASM project - it just sseemed like a good idea to add a standalone command line version, and now a WebGPU 
implementation for better interactivity.

[<img width="256" src="gallery/cli/acoustic.png">](gallery/cli/acoustic.png)
[<img width="256" src="gallery/webgpu/line-bw.png">](gallery/webgpu/line-bw.png)
[<img width="256" src="gallery/webgpu/gradient-yellow.png">](gallery/webgpu/gradient-yellow.png)
[<img width="256" src="gallery/webgpu/gradient3E.png">](gallery/webgpu/gradient3E.png)

## Releases

| *Version* | *Description*                                                                                            |
| --------- | -------------------------------------------------------------------------------------------------------- |
| v1.1.0    | Added `wav2mp4` command line utility                                                                     |
| v1.0.0    | Initial release                                                                                          |


## Installation

Platform specific executables can be downloaded from the [releases](https://github.com/transcriptaze/wav2png/releases) 
page. Installation is straightforward - download and extract the archive for your platform and place the executables in 
a directory in your PATH. 

### Building from source

#### Command line
Assuming you have `Go` and `make` installed:

```
git clone https://github.com/transcriptaze/wav2png.git
cd wav2png/go
make build
```

If you prefer not to use `make`:
```
git clone https://github.com/transcriptaze/wav2png.git
cd wav2png/go
mkdir bin
go build -o bin ./...
```

#### WebGPU

## Usage

### wav2png

Command line:
```
wav2png [--debug] [options] [--out <path>] <wav>

  <wav>         WAV file to render.

  --out <path>  File path for PNG file - if <path> is a directory, the WAV file name is
                used. Defaults to the WAV file base path.

  --debug       Displays occasionally useful diagnostic information.

Options:
  --settings <file>      JSON file with the default settings for the height, width, etc. Defaults to 
                         .settings.json if not specified, falling back to internal default values if 
                         .settings.json does not exist.

  --width <pixels>       Width (in pixels) of the PNG image. Valid values are in the range 32 to 8192, 
                         defaults to 645px.

  --height <pixels>      Height (in pixels) of the PNG image. Valid values are in the range 32 to 8192, 
                         defaults to 395px.
  
  --padding <pixels>     Padding (in pixels) between the border of the PNG and the extent of the rendered
                         waveform. Valid values are -16 to +32, defaults to 2px.

  --palette <palette>    Palette used to colour the waveform. May be the name of one of the internal colour
                         palettes or a user provided PNG file. Defaults to 'ice'
  
  --fill <fillspec>      Fill specification for the background colour, in the form type:colour 
                         e.g. solid:#0000ffff. Currently the only fill type supported is 'solid', defaults
                         to solid:#000000ff.

  --grid <gridspec>      Grid specification for an optional rectilinear grid, in the form 
                         type:colour:size:overlay, e.g.
                         - none
                         - square:#008000ff:~64
                         - rectangle:#008000ff:~64x48:overlay
                         
                         The size may preceded by a 'fit':
                         - ~  approximate
                         - =  exact
                         - ≥  at least
                         - ≤  at most
                         - >  greater than
                         - <  less than

                         If gridspec includes :overlay, the grid is rendered 'in front' of the waveform.

                         The default gridspec is 'square:#008000ff:~64'

  --antialias <kernel>   The antialising kernel applied to soften the rendered PNG. Valid values are:
                         - none        no antialiasing
                         - horizontal  blurs horizontal edges
                         - vertical    blurs vertical edges
                         - soft        blurs both horizontal and vertical edges

                         The default kernel is 'vertical'.

  --scale <scale>        A vertical scaling factor to size the height of the rendered waveform. The valid 
                         range is 0.2 to 5.0, defaults to 1.0.

  --mix  <mixspec>       Specifies how to combine channels from a stereo WAV file. Valid values are:
                         - 'L'    Renders the left channel only
                         - 'R'    Renders the right channel only
                         - 'L+R'  Combines the left and right channels
                         
                         Defaults to 'L+R'.

  --start <time>         The start time of the segment of audio to render, in Go time format (e.g. 10s or
                         1m5s). Defaults to 0s.

  --end <time>           The end time of the segment of audio to render, in Go time format (e.g. 10s or
                         1m5s). Defaults to the end of the audio.


Example:

wav2png --debug                                        \
          --settings 'settings.json'                     \
          --height 390                                   \
          --width 641                                    \
          --padding 0                                    \
          --palette 'amber.png'                          \
          --fill 'solid:#0000ffff'                       \
          --grid 'rectangular:#800000ff:~32x128:overlay' \
          --antialias 'soft'                             \
          --scale 0.5                                    \
          --start 0.5s                                   \
          --end 1.5s                                     \
          --mix 'L+R'                                    \
          --out example.png                              \
          example.wav
```

## wav2mp4

Command line:
```
wav2mp4 [--debug] [options] [--out <path>] --window <duration> --fps <frame rate> --cursor <cursor spec> <wav>

  <wav>                  WAV file to render.

  --out <path>           File path for MP4 file - if <path> is a directory, the WAV file name is
                         used and defaults to the WAV file base path. wav2mp4 generates a set of ffmpeg
                         frames files in the 'frames' subdirectory of the out file directory. 

  --window <duration>    The interval allotted to a single frame, in Go time format e.g. --window 1s.
                         The window interval must be less than the duration of the MP4.

  --fps <frame rate>     Frame rate for the MP4 in frames per second e.g. --fps 30

  --cursor <cursorspec>  Cursor to indicate for the current play position. A cursor is specified by the 
                         image source and dynamic:

                         --cursor <image>:<dynamic>

                         where image may be:
                         - none
                         - red  (internal 'red' cursor)
                         - blue (internal 'blue' cursor)
                         - a PNG file with a custom cursor image

                         The cursor 'dynamic' defaults to 'sweep' if not specified, but may be one of
                         the following:
                         - sweep  Moves linearly from left to right over the duration of the MP4
                         - left   Fixed on left side
                         - right  Fixed on right side
                         - center Fixed in center of frame
                         - ease   Migrates from the left to center of the frame, before moving to the
                                  right side to finish
                         - erf    Moves 'sigmoidally' from left to right over the duration of the MP4, 
                                  with the sigmoid defined by the inverse error function

  --debug                Displays occasionally useful diagnostic information.

Options:
  --settings <file>      JSON file with the default settings for the height, width, etc. Defaults to 
                         .settings.json if not specified, falling back to internal default values if 
                         .settings.json does not exist.

  --width <pixels>       Width (in pixels) of the PNG image. Valid values are in the range 32 to 8192, 
                         defaults to 645px.

  --height <pixels>      Height (in pixels) of the PNG image. Valid values are in the range 32 to 8192, 
                         defaults to 395px.
  
  --padding <pixels>     Padding (in pixels) between the border of the PNG and the extent of the 
                         rendered waveform. Valid values are -16 to +32, defaults to 2px.

  --palette <palette>    Palette used to colour the waveform. May be the name of one of the internal 
                         colour palettes or a user provided PNG file. Defaults to 'ice'
  
  --fill <fillspec>      Fill specification for the background colour, in the form type:colour 
                         e.g. solid:#0000ffff. Currently the only fill type supported is 'solid', 
                         defaults to solid:#000000ff.

  --grid <gridspec>      Grid specification for an optional rectilinear grid, in the form 
                         type:colour:size:overlay, e.g.
                         - none
                         - square:#008000ff:~64
                         - rectangle:#008000ff:~64x48:overlay
                         
                         The size may preceded by a 'fit':
                         - ~  approximate
                         - =  exact
                         - ≥  at least
                         - ≤  at most
                         - >  greater than
                         - <  less than

                         If gridspec includes :overlay, the grid is rendered 'in front' of the 
                         waveform.

                         The default gridspec is 'square:#008000ff:~64'

  --antialias <kernel>   The antialising kernel applied to soften the rendered PNG. Valid values are:
                         - none        no antialiasing
                         - horizontal  blurs horizontal edges
                         - vertical    blurs vertical edges
                         - soft        blurs both horizontal and vertical edges

                         The default kernel is 'vertical'.

  --scale <scale>        A vertical scaling factor to size the height of the rendered waveform. The 
                         valid range is 0.2 to 5.0, defaults to 1.0.

  --mix  <mixspec>       Specifies how to combine channels from a stereo WAV file. Valid values are:
                         - 'L'    Renders the left channel only
                         - 'R'    Renders the right channel only
                         - 'L+R'  Combines the left and right channels
                         
                         Defaults to 'L+R'.

  --start <time>         The start time of the segment of audio to render, in Go time format (e.g. 10s
                         or 1m5s). Defaults to 0s.

  --end <time>           The end time of the segment of audio to render, in Go time format (e.g. 10s
                         or 1m5s). Defaults to the end of the audio.


Example:

  wav2mp4 --debug --window 1s --fps 30 --cursor red:sweep --out ./example/chirp.mp4 ./samples/chirp.wav
  cd ./example/frames
  ffmpeg -framerate 30 -i frame-%05d.png -c:v libx264 -pix_fmt yuv420p -crf 23 -y out.mp4
  ffmpeg -i out.mp4    -i ./samples/chirp.wav -c:v copy -c:a aac -y ../chirp.mp4
```

## References & Related Projects

1. [Audio File Format Specifications](http://www-mmsp.ece.mcgill.ca/Documents/AudioFormats/WAVE/WAVE.html)
2. [SoX](http://sox.sourceforge.net)
3. [WaveFile Gem](https://wavefilegem.com/how_wave_files_work.html)
4. [DirectMusic: wav2png](https://directmusic.me/wav2png)
5. [Shulz Audio: wav2png](https://schulz.audio/products/wav2png)
6. [iconmonstr](https://iconmonstr.com/sound-wave-2-png)
7. [BBC](https://github.com/bbc/audiowaveform)
8. [BBC CLI](https://github.com/marc7806/bbc-audiowaveform-cli-wrapper)
9. [stackoverflow](https://stackoverflow.com/questions/4468157/how-can-i-create-a-waveform-image-of-an-mp3-in-linux)
10. [github:waveform](https://github.com/andrewrk/waveform)


CHANGELOG.md
# CHANGELOG

## Unreleased

### Added
1. Columns renderer
2. WebGPU implementation

### Updated
1. Moved Go renderer to _github.com/transcriptaze/wav2png/go_ package.


## [1.1.0](https://github.com/transcriptaze/wav2png/releases/tag/v1.0.0) - 2021-07-13

### Added
1. Added _wav2mp4_ command line utility to generate an ffmpeg frame set.


## [1.0.0](https://github.com/transcriptaze/wav2png/releases/tag/v1.0.0) - 2021-06-25

1. Initial release.



styles/grid.go
package styles

import (
	"fmt"
	"image/color"
	"regexp"
	"strconv"
	"strings"

	"github.com/transcriptaze/wav2png/grids"
)

type Grid struct {
	Grid    string `json:"grid"`
	Colour  string `json:"colour"`
	Alpha   uint8  `json:"alpha"`
	Size    string `json:"size"`
	WH      string `json:"wh"`
	Overlay bool   `json:"overlay"`
}

func (g Grid) String() string {
	if g.Grid == "none" {
		return fmt.Sprintf("%v", g.Grid)
	}

	if g.Grid == "square" {
		if g.Overlay {
			return fmt.Sprintf("%v:%v%02x:%v:overlay", g.Grid, g.Colour, g.Alpha, g.Size)
		} else {
			return fmt.Sprintf("%v:%v%02x:%v", g.Grid, g.Colour, g.Alpha, g.Size)
		}
	}

	if g.Grid == "rectangular" {
		if g.Overlay {
			return fmt.Sprintf("%v:%v%02x:%v:overlay", g.Grid, g.Colour, g.Alpha, g.WH)
		} else {
			return fmt.Sprintf("%v:%v%02x:%v", g.Grid, g.Colour, g.Alpha, g.WH)
		}
	}

	return "??"
}

func (g *Grid) Set(s string) error {
	ss := strings.ToLower(s)
	match := regexp.MustCompile("^(none|square|rectangular).*").FindStringSubmatch(ss)

	if len(match) > 1 {
		switch match[1] {
		case "none":
			g.Grid = "none"

		case "square":
			g.Grid = "square"

			match = regexp.MustCompile("^square:(#[[:xdigit:]]{8}).*").FindStringSubmatch(ss)
			if len(match) > 1 {
				color := colour(match[1])
				g.Colour = fmt.Sprintf("#%02x%02x%02x", color.R, color.G, color.B)
				g.Alpha = color.A
			}

			match = regexp.MustCompile("^square:#[[:xdigit:]]{8}:([~=≥≤><]?[0-9]+).*").FindStringSubmatch(ss)
			if len(match) > 1 {
				fit, size := size(match[1])
				g.Size = fmt.Sprintf("%v%v", fit, size)
			}

			match = regexp.MustCompile("^square:#[[:xdigit:]]{8}:[~=≥≤><]?[0-9]+:(overlay)").FindStringSubmatch(ss)
			if len(match) > 1 {
				g.Overlay = true
			}

		case "rectangular":
			g.Grid = "rectangular"

			match = regexp.MustCompile("^rectangular:(#[[:xdigit:]]{8}).*").FindStringSubmatch(ss)
			if len(match) > 1 {
				color := colour(match[1])
				g.Colour = fmt.Sprintf("#%02x%02x%02x", color.R, color.G, color.B)
				g.Alpha = color.A
			}

			match = regexp.MustCompile("^rectangular:#[[:xdigit:]]{8}:([~=≥≤><]?[0-9]+x[0-9]+).*").FindStringSubmatch(ss)
			if len(match) > 1 {
				fit, w, h := wh(match[1])
				g.WH = fmt.Sprintf("%v%vx%v", fit, w, h)
			}

			match = regexp.MustCompile("^rectangular:#[[:xdigit:]]{8}:[~=≥≤><]?[0-9]+x[0-9]+:(overlay)").FindStringSubmatch(ss)
			if len(match) > 1 {
				g.Overlay = true
			}

		}
	}

	return nil
}

func (g Grid) GridSpec() grids.GridSpec {
	// ... overlay
	overlay := g.Overlay

	// ... colour
	red := uint8(0)
	green := uint8(128)
	blue := uint8(0)
	alpha := g.Alpha

	colour := color.NRGBA{R: red, G: green, B: blue, A: alpha}

	if _, err := fmt.Sscanf(g.Colour, "#%02x%02x%02x", &red, &green, &blue); err == nil {
		colour = color.NRGBA{R: red, G: green, B: blue, A: g.Alpha}
	}

	// ... no grid
	if g.Grid == "none" {
		return grids.NewNoGrid()
	}

	// ... rectangular
	if g.Grid == "rectangular" {
		fit := grids.Approximate
		width := uint(64)
		height := uint(48)

		if matched := regexp.MustCompile(`([~=><≥≤])?\s*([0-9]+)x([0-9]+)`).FindStringSubmatch(g.WH); len(matched) == 4 {
			switch matched[1] {
			case "~":
				fit = grids.Approximate
			case "=":
				fit = grids.Exact
			case "≥":
				fit = grids.AtLeast
			case "≤":
				fit = grids.AtMost
			case ">":
				fit = grids.LargerThan
			case "<":
				fit = grids.SmallerThan
			}

			if v, err := strconv.ParseUint(matched[2], 10, 32); err == nil && v >= 16 && v <= 1024 {
				width = uint(v)
			}

			if v, err := strconv.ParseUint(matched[3], 10, 32); err == nil && v >= 16 && v <= 1024 {
				height = uint(v)
			}
		}

		return grids.NewRectangularGrid(colour, width, height, fit, overlay)
	}

	// ... default to square
	fit := grids.Approximate
	size := uint(64)

	if matched := regexp.MustCompile(`([~=><≥≤])?\s*([0-9]+)`).FindStringSubmatch(g.Size); len(matched) == 3 {
		switch matched[1] {
		case "~":
			fit = grids.Approximate
		case "=":
			fit = grids.Exact
		case "≥":
			fit = grids.AtLeast
		case "≤":
			fit = grids.AtMost
		case ">":
			fit = grids.LargerThan
		case "<":
			fit = grids.SmallerThan
		}

		if v, err := strconv.ParseUint(matched[2], 10, 32); err == nil && v >= 16 && v <= 1024 {
			size = uint(v)
		}
	}

	return grids.NewSquareGrid(colour, size, fit, overlay)
}

func size(s string) (grids.Fit, uint) {
	fit := grids.Approximate
	size := uint(64)

	if matched := regexp.MustCompile(`([~=><≥≤])?\s*([0-9]+)`).FindStringSubmatch(s); len(matched) == 3 {
		switch matched[1] {
		case "~":
			fit = grids.Approximate
		case "=":
			fit = grids.Exact
		case "≥":
			fit = grids.AtLeast
		case "≤":
			fit = grids.AtMost
		case ">":
			fit = grids.LargerThan
		case "<":
			fit = grids.SmallerThan
		}

		if v, err := strconv.ParseUint(matched[2], 10, 32); err == nil && v >= 16 && v <= 1024 {
			size = uint(v)
		}
	}

	return fit, size
}

func wh(s string) (grids.Fit, uint, uint) {
	fit := grids.Approximate
	width := uint(64)
	height := uint(48)

	if matched := regexp.MustCompile(`([~=><≥≤])?\s*([0-9]+)x([0-9]+)`).FindStringSubmatch(s); len(matched) == 4 {
		switch matched[1] {
		case "~":
			fit = grids.Approximate
		case "=":
			fit = grids.Exact
		case "≥":
			fit = grids.AtLeast
		case "≤":
			fit = grids.AtMost
		case ">":
			fit = grids.LargerThan
		case "<":
			fit = grids.SmallerThan
		}

		if v, err := strconv.ParseUint(matched[2], 10, 32); err == nil && v >= 16 && v <= 1024 {
			width = uint(v)
		}

		if v, err := strconv.ParseUint(matched[3], 10, 32); err == nil && v >= 16 && v <= 1024 {
			height = uint(v)
		}
	}

	return fit, width, height
}
